SPARK-1706: Allow multiple executors per worker in Standalone mode

CodingCat opened the pull request:
resubmit of https://github.com/apache/spark/pull/636  for a totally different algorithm

https://issues.apache.org/jira/browse/SPARK-1706

In current implementation, the user has to start multiple workers in a server for starting multiple executors in a server, which introduces additional overhead due to the more JVM processes...

In this patch, I changed the scheduling logic in master to enable the user to start multiple executor processes within the same JVM process.
1. user configure spark.executor.maxCoreNumPerExecutor to suggest the maximum core he/she would like to allocate to each executor
2. Master assigns the executors to the workers with the major consideration on the memoryPerExecutor and the worker.freeMemory, and tries to allocate as many as possible cores to the executor `min(min(memoryPerExecutor, worker.freeCore), maxLeftCoreToAssign)` where `maxLeftCoreToAssign = maxExecutorCanAssign * maxCoreNumPerExecutor`

---

Other small changes include

change memoryPerSlave in ApplicationDescription to memoryPerExecutor, as "Slave" is overrided to represent both worker and executor in the documents... (we have some discussion on this before?)

CodingCat committed:
make master support multiple executors per worker

nchammas commented:
@CodingCat Do we need to ping anyone specific to look at this PR? It's been many months since the last update.

CodingCat commented:
I will rebase this one and send email to Kay....

SparkQA commented:
[Test build #25251 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25251/consoleFull) for   PR 731 at commit [`b4a8a68`](https://github.com/apache/spark/commit/b4a8a68c26bda2b272ead13bb747c75f19be9081).
 * This patch merges cleanly.

CodingCat commented:
@nchammas , sorry, thought it was SPARK-1143..I think it is supposed to be reviewed by @andrewor14 @pwendell and @mridulm ???

SparkQA commented:
[Test build #25251 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25251/consoleFull) for   PR 731 at commit [`b4a8a68`](https://github.com/apache/spark/commit/b4a8a68c26bda2b272ead13bb747c75f19be9081).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25251/
Test FAILed.

SparkQA commented:
[Test build #25266 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25266/consoleFull) for   PR 731 at commit [`782dfcb`](https://github.com/apache/spark/commit/782dfcb815b4ab8935444f69457727fed28baa34).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25266 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25266/consoleFull) for   PR 731 at commit [`782dfcb`](https://github.com/apache/spark/commit/782dfcb815b4ab8935444f69457727fed28baa34).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25266/
Test FAILed.

SparkQA commented:
[Test build #25323 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25323/consoleFull) for   PR 731 at commit [`2608d11`](https://github.com/apache/spark/commit/2608d1130fc7bc497ef09f7a08cb47e10d0829c7).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25323 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25323/consoleFull) for   PR 731 at commit [`2608d11`](https://github.com/apache/spark/commit/2608d1130fc7bc497ef09f7a08cb47e10d0829c7).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25323/
Test PASSed.

SparkQA commented:
[Test build #25853 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25853/consoleFull) for   PR 731 at commit [`d2f413a`](https://github.com/apache/spark/commit/d2f413a5e7eebc680aff5942e841ace0118612ed).
 * This patch merges cleanly.

SparkQA commented:
[Test build #25853 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/25853/consoleFull) for   PR 731 at commit [`d2f413a`](https://github.com/apache/spark/commit/d2f413a5e7eebc680aff5942e841ace0118612ed).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/25853/
Test PASSed.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -581,6 +552,99 @@ private[spark] class Master(
     }
   }
 
+  private def startMultiExecutorsPerWorker() {
```
This method definitely warrants a large java doc to explain what it's doing...

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -502,43 +502,14 @@ private[spark] class Master(
    * launched an executor for the app on it (right now the standalone backend doesn't like having
    * two executors on the same worker).
```
this comment is outdated

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -502,43 +502,14 @@ private[spark] class Master(
-    // Right now this is a very simple FIFO scheduler. We keep trying to fit in the first app
-    // in the queue, then the second app, etc.
+  // Right now this is a very simple FIFO scheduler. We keep trying to fit in the first app
+  // in the queue, then the second app, etc.
```
this should be a java doc

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -502,43 +502,14 @@ private[spark] class Master(
-    // in the queue, then the second app, etc.
+  // Right now this is a very simple FIFO scheduler. We keep trying to fit in the first app
+  // in the queue, then the second app, etc.
+  private def startSingleExecutorPerWorker() {
```
This method should have a java doc not a comment

andrewor14 commented:
@CodingCat Thanks for doing this and sorry for the delay. I haven't dug into the detail of `startMultiExecutorPerWorker`. Also, I don't see why we need a separate config to enable this. If we don't set the max cores per executor, then I think it's OK to assume that the executor will take all the cores (i.e. the max cores per executor defaults to Int.MaxValue or something, which would be the same as old behavior). If we do this, then do we still need to differentiate `startMultiExecutorPerWorker` and `startSingleExecutorPerWorker`?

CodingCat committed:
java doc

SparkQA commented:
[Test build #26162 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26162/consoleFull) for   PR 731 at commit [`cfb7d23`](https://github.com/apache/spark/commit/cfb7d23d805afa780f071f6cbd7bf8efcd223cc3).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26162 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26162/consoleFull) for   PR 731 at commit [`cfb7d23`](https://github.com/apache/spark/commit/cfb7d23d805afa780f071f6cbd7bf8efcd223cc3).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26162/
Test FAILed.

CodingCat committed:
stylistic fix

SparkQA commented:
[Test build #26163 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26163/consoleFull) for   PR 731 at commit [`1808d3a`](https://github.com/apache/spark/commit/1808d3a9a78f6016cba2678612607893662f82d7).
 * This patch merges cleanly.

CodingCat committed:
more java docs and consolidate canUse function

SparkQA commented:
[Test build #26164 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26164/consoleFull) for   PR 731 at commit [`9f0c3a4`](https://github.com/apache/spark/commit/9f0c3a4393933e77c5e97a322d7bd9038afc7f78).
 * This patch merges cleanly.

CodingCat committed:
test commit

SparkQA commented:
[Test build #26165 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26165/consoleFull) for   PR 731 at commit [`97918d2`](https://github.com/apache/spark/commit/97918d2753359881dfd7f512bedc4495e47d3599).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26163 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26163/consoleFull) for   PR 731 at commit [`1808d3a`](https://github.com/apache/spark/commit/1808d3a9a78f6016cba2678612607893662f82d7).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26163/
Test PASSed.

SparkQA commented:
[Test build #26165 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26165/consoleFull) for   PR 731 at commit [`97918d2`](https://github.com/apache/spark/commit/97918d2753359881dfd7f512bedc4495e47d3599).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DenseMatrix(`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26165/
Test FAILed.

SparkQA commented:
[Test build #26164 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26164/consoleFull) for   PR 731 at commit [`9f0c3a4`](https://github.com/apache/spark/commit/9f0c3a4393933e77c5e97a322d7bd9038afc7f78).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DenseMatrix(`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26164/
Test PASSed.

CodingCat committed:
remove outdated comments

SparkQA commented:
[Test build #26166 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26166/consoleFull) for   PR 731 at commit [`258d4b5`](https://github.com/apache/spark/commit/258d4b540319217d367173949e41dc912b8663bc).
 * This patch merges cleanly.

CodingCat commented:
Hi, @andrewor14 , thanks for the review, I just addressed your comments except the one about merging startMultiExecutor/startSingleExecutor

I think the key difference here is that, we want to make start multiple executors per worker as an application specific option, or a global option (like spreadApp)

if the former one, I guess I can find some way to merge these two functions . if the latter one, we truly need an additional spark.* system properties passed to master's SparkConf, since in schedule() function we cannot get the application specific "spark.executor.maxCoreNumPerExecutor"... your thoughts? (I guess I add to rename the configuration entries to spark.deploy.*...)

CodingCat commented:
oh, before you merge, please drop me a message to let me disable this feature by default (now I enabled it just for testing)

SparkQA commented:
[Test build #26166 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26166/consoleFull) for   PR 731 at commit [`258d4b5`](https://github.com/apache/spark/commit/258d4b540319217d367173949e41dc912b8663bc).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26166/
Test PASSed.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/ApplicationDescription.scala:
```diff
@@ -20,13 +20,14 @@ package org.apache.spark.deploy
   val user = System.getProperty("user.name", "<unknown>")
-
+  // only valid when spark.executor.multiPerWorker is set to true
+  var maxCorePerExecutor = maxCores
```
this setting doesn't make sense. `maxCores` is the number of cores shared across all nodes in the cluster, not the number of cores on each executor.

andrewor14 commented on docs/configuration.md:
```diff
@@ -1207,6 +1207,25 @@ Apart from these, the following properties are also available, and may be useful
+  </td>
+</tr>
+<tr>
+  <td>spark.executor.maxCoreNumPerExecutor</td>
```
I agree with @nishkamravi2's suggestion to just reuse `spark.executor.cores` here. No reason to introduce another hard-to-name config that we can't remove in the future.

andrewor14 commented on docs/configuration.md:
```diff
@@ -1207,6 +1207,25 @@ Apart from these, the following properties are also available, and may be useful
+  </td>
+</tr>
+<tr>
+  <td>spark.executor.extraClassPath</td>
```
wait, what is this doing here? The description doesn't match this config. Did you add this by mistake?

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -581,6 +558,104 @@ private[spark] class Master(
+        var maxCoresLeft = app.coresLeft
+        val numUsable = usableWorkers.length
+        // 2D array to track the number of cores of each executor assigned to each worker
+        val assigned = Array.fill[ListBuffer[Int]](numUsable)(new ListBuffer[Int])
```
we should explain in more detail what this 2D array contains:
```
// A 2D array that tracks the number of cores used by each executor launched on
// each worker. The first index refers to the usable worker, and the second index
// refers to the executor launched on that worker.
```

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/ui/ApplicationPage.scala:
```diff
@@ -74,18 +74,18 @@ private[spark] class ApplicationPage(parent: MasterWebUI) extends WebUIPage("app
                   app.desc.maxCores.get, app.coresGranted, app.coresLeft)
               }
-            }
+              }
```
let's revert these indentation changes

andrewor14 commented:
@CodingCat I looked at this patch much more closely, and I still don't see a need to separate the single and the multiple executors per worker cases. More specifically, I see the single executor case as a special case of the multiple executors case, where each element of your 2D array will have a list of 1 element (because there's only one executor).

I think it makes more sense to configure the number of executors per worker directly. Perhaps we need a config that looks something like `spark.deploy.executorsPerWorker`. Then, to avoid one executor from grabbing all the cores on that worker, the user will also need to set `spark.executor.cores`. In fact, we're doing something fairly similar in Mesos coarse-grained mode in this PR #4027. It would be good to model the general structure of the changes here after the one there.

sryza commented on docs/configuration.md:
```diff
@@ -1207,6 +1207,25 @@ Apart from these, the following properties are also available, and may be useful
   <td>Empty</td>
+</tr>
+<tr>
+  <td>spark.executor.multiPerWorker</td>
```
This is fundamentally a worker property, not an executor property, as it's the worker that reads it and makes decisions based on it, right?  If that's the case, this shouldn't be under the `spark.executor` namespace.

CodingCat committed:
address the comments

SparkQA commented:
[Test build #27836 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27836/consoleFull) for   PR 731 at commit [`f4abf0c`](https://github.com/apache/spark/commit/f4abf0c9b24b73b05642ee330d95e2743bac1b16).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27836 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27836/consoleFull) for   PR 731 at commit [`f4abf0c`](https://github.com/apache/spark/commit/f4abf0c9b24b73b05642ee330d95e2743bac1b16).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27836/
Test FAILed.

SparkQA commented:
[Test build #27838 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27838/consoleFull) for   PR 731 at commit [`7fd8f7b`](https://github.com/apache/spark/commit/7fd8f7b93c664490b726483b402ba953749e4ac4).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27838 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27838/consoleFull) for   PR 731 at commit [`7fd8f7b`](https://github.com/apache/spark/commit/7fd8f7b93c664490b726483b402ba953749e4ac4).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27838/
Test FAILed.

CodingCat committed:
fix compilation issue

SparkQA commented:
[Test build #27839 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27839/consoleFull) for   PR 731 at commit [`5f90cd1`](https://github.com/apache/spark/commit/5f90cd1021b68a9ceb93c9412e84c76ce193df2d).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27839 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27839/consoleFull) for   PR 731 at commit [`5f90cd1`](https://github.com/apache/spark/commit/5f90cd1021b68a9ceb93c9412e84c76ce193df2d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27839/
Test PASSed.

CodingCat commented:
![image](https://cloud.githubusercontent.com/assets/678008/6320817/b3d4efb2-bab6-11e4-9615-1530d981e286.png)


![image](https://cloud.githubusercontent.com/assets/678008/6320824/c21b00fc-bab6-11e4-8b70-f6c945052527.png)

CodingCat committed:
address Andrew's comments

SparkQA commented:
[Test build #27842 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27842/consoleFull) for   PR 731 at commit [`d6c19b4`](https://github.com/apache/spark/commit/d6c19b42e29944f0d6a0ff3c32741e0ce6a9a596).
 * This patch merges cleanly.

CodingCat commented:
Hi, @andrewor14 , thanks for your review 

I think I misunderstood your original point. Now, I merged two functions (starting one/more executor(s)) .

The difference between my current approach and your proposal is that I didn't introduce ###### spark.deploy.executorsPerWorker, instead, the executor number on each worker is determined by the available cores/memory of the worker at the scheduling moment.


The snapshots of running 4 executors in 2 workers were attached in my last comment.

SparkQA commented:
**[Test build #27842 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27842/consoleFull)**     for PR 731 at commit [`d6c19b4`](https://github.com/apache/spark/commit/d6c19b42e29944f0d6a0ff3c32741e0ce6a9a596)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27842/
Test FAILed.

CodingCat committed:
bug fix

SparkQA commented:
[Test build #27843 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27843/consoleFull) for   PR 731 at commit [`d941fa9`](https://github.com/apache/spark/commit/d941fa916b0b77ee00bc4f81fd6157193ed89b9d).
 * This patch merges cleanly.

CodingCat committed:
typo fix

SparkQA commented:
[Test build #27844 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27844/consoleFull) for   PR 731 at commit [`97ed489`](https://github.com/apache/spark/commit/97ed489597ef3599f93a1a06fa889e1536410c88).
 * This patch merges cleanly.

CodingCat committed:
change the launching executor code

SparkQA commented:
[Test build #27845 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27845/consoleFull) for   PR 731 at commit [`2c6d26a`](https://github.com/apache/spark/commit/2c6d26a9ba18c628c6a6d0b471d7be01c07dddb1).
 * This patch merges cleanly.

SparkQA commented:
**[Test build #27843 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27843/consoleFull)**     for PR 731 at commit [`d941fa9`](https://github.com/apache/spark/commit/d941fa916b0b77ee00bc4f81fd6157193ed89b9d)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27843/
Test FAILed.

SparkQA commented:
**[Test build #27844 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27844/consoleFull)**     for PR 731 at commit [`97ed489`](https://github.com/apache/spark/commit/97ed489597ef3599f93a1a06fa889e1536410c88)     after a configured wait of `120m`.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27844/
Test FAILed.

SparkQA commented:
[Test build #27845 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27845/consoleFull) for   PR 731 at commit [`2c6d26a`](https://github.com/apache/spark/commit/2c6d26a9ba18c628c6a6d0b471d7be01c07dddb1).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27845/
Test PASSed.

Du-Li commented:
Has this PR been merged/released? It's a very useful patch. Thanks for working it out.

CodingCat commented:
@Du-Li , thanks for looking at the patch

This patch is still under review, we will get it merged ASAP

SparkQA commented:
[Test build #28302 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28302/consoleFull) for   PR 731 at commit [`dd39148`](https://github.com/apache/spark/commit/dd391484356e3af490a8aa61dbe1b06e2db10d56).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28302 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28302/consoleFull) for   PR 731 at commit [`dd39148`](https://github.com/apache/spark/commit/dd391484356e3af490a8aa61dbe1b06e2db10d56).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28302/
Test PASSed.

CodingCat commented:
@andrewor14 would you mind taking further review, when you have time?

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/ApplicationDescription.scala:
```diff
@@ -41,5 +41,8 @@ private[spark] class ApplicationDescription(
       name, maxCores, memoryPerSlave, command, appUiUrl, eventLogDir, eventLogCodec)
 
+  // only valid when spark.executor.multiPerWorker is set to true
+  var maxCorePerExecutor: Option[Int] = None
```
This shouldn't be a `var` because we only ever set it in one place (when we create it). This should just go into the constructor.

CodingCat replied:
but will MIMA be happy about it?

andrewor14 replied:
this is `private[spark] class` so it won't matter. Also it should be `maxCoresPerExecutor`.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -351,6 +351,8 @@ object SparkSubmit {
       OptionAssigner(args.driverMemory, STANDALONE, CLUSTER, sysProp = "spark.driver.memory"),
       OptionAssigner(args.driverCores, STANDALONE, CLUSTER, sysProp = "spark.driver.cores"),
+      OptionAssigner(args.executorCores, STANDALONE, ALL_DEPLOY_MODES, 
+        sysProp = "spark.executor.cores"),
```
this is in the wrong place because it's not just standalone cluster mode. It should be under `// Other options` in L378

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala:
```diff
@@ -498,11 +498,18 @@ private[spark] class SparkSubmitArguments(args: Seq[String], env: Map[String, St
+        |                              than 1 executors on each worker in standalone mode; 
+        |                              otherwise, only one executor on each executor is allowed 
+        |                              (the executor will take all available cores of the worker at 
+        |                              the moment.
```
I don't understand what this is saying. Maybe you mean
```
Number of cores to use on each executor. In standalone mode,
the executor will use all available cores on the worker if this
is not specified. (Default: 1 in YARN mode, all available cores
in standalone mode)
```

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -513,89 +513,122 @@ private[spark] class Master(
+   * Schedule the currently available resources among waiting apps. This method will be called
+   * every time a new app joins or resource availability changes.
+   */
+  def schedule() {
```
`private def scheduler(): Unit = {`

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -513,89 +513,122 @@ private[spark] class Master(
+  private def canUse(app: ApplicationInfo, worker: WorkerInfo): Boolean = {
+    val enoughResources = worker.memoryFree >= app.desc.memoryPerExecutorMB && worker.coresFree > 0
+    val allowToExecute = app.desc.maxCorePerExecutor.isDefined || !worker.hasExecutor(app)
+    allowToExecute && enoughResources
```
I'm not sure if we need `allowToExecute` here. If `maxCorePerExecutor` is not defined then we will grab all the cores, in which case `worker.coresFree` will be 0.

andrewor14 commented:
@CodingCat Thanks for merging the two cases. However, I still find the code overly complex. In particular, I think we can modify far less by rewriting [this section](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/deploy/master/Master.scala#L578) of the original code with something like the following:
```
// Now that we've decided how many cores to give on each node, let's actually give them
for (pos <- 0 until numUsable) {
  while (assigned(pos) > 0) {
    val maxCoresPerExecutor = app.maxCoresPerExecutor.getOrElse(Int.MaxValue)
    val coresForThisExecutor = math.min(maxCoresPerExecutor, assigned(pos))
    val exec = app.addExecutor(usableWorkers(pos), coresForThisExecutor)
    assigned(pos) -= coresForThisExecutor
    launchExecutor(usableWorkers(pos), exec)
    app.state = ApplicationState.RUNNING
  }
}
```
then in the else case (where we don't spread out apps), we have to do something similar.

Another potential issue is that if we schedule multiple executors that belong to the same app on the same worker, we might not be able to distinguish them. This is what Patrick was suggesting on the [JIRA](https://issues.apache.org/jira/browse/SPARK-1706) (see point 3). I haven't verified whether this is actually a problem, but it would be good to keep it in mind.

andrewor14 commented:
Also, while I was reviewing this I had more thoughts on the appropriate config to use. Given the semantics that we want to support, I actually think it may not be correct to reuse `spark.executor.cores`. The reason is that here all we guarantee is that the executor gets *at most*, not exactly, `spark.executor.cores`. This is inconsistent with the semantics of this config in YARN.

Instead, maybe it makes sense to configure this in a new config: `spark.deploy.maxCoresPerExecutor`.

CodingCat commented:
@andrewor14 thanks for your patient review, I will address them on Thu. or Fri.

CodingCat committed:
address andrew's comments

SparkQA commented:
[Test build #29269 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29269/consoleFull) for   PR 731 at commit [`f74cd8e`](https://github.com/apache/spark/commit/f74cd8e29368add06e6fc960cf11988f25d5418e).
 * This patch merges cleanly.

CodingCat commented:
@andrewor14 I think in spreadApp mode, grabing all cores  when `spark.deploy.maxCoresPerExecutor` is not defined is *not* the right approach...

if we grab all cores, it is exactly equivalent to `non-spreadApp mode` in the implementation in current master branch. Instead, we should traverse all workers one by one, for each visit, we allocate 1 free core to the executor when  `spark.deploy.maxCoresPerExecutor == None`; otherwise, for each visit, we assign at most `spark.deploy.maxCoresPerExecutor`  cores


am I right?

CodingCat committed:
fix wrong usage info

CodingCat commented:
snapshot

![image](https://cloud.githubusercontent.com/assets/678008/6860808/99e0f450-d403-11e4-8ac0-2191640c0412.png)

SparkQA commented:
[Test build #29270 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29270/consoleFull) for   PR 731 at commit [`193d687`](https://github.com/apache/spark/commit/193d6873a33cd8582c329e62d34c5fe52d5b6c1e).
 * This patch merges cleanly.

SparkQA commented:
[Test build #29269 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29269/consoleFull) for   PR 731 at commit [`f74cd8e`](https://github.com/apache/spark/commit/f74cd8e29368add06e6fc960cf11988f25d5418e).
 * This patch **passes all tests**.

 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29269/
Test PASSed.

SparkQA commented:
[Test build #29270 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29270/consoleFull) for   PR 731 at commit [`193d687`](https://github.com/apache/spark/commit/193d6873a33cd8582c329e62d34c5fe52d5b6c1e).
 * This patch **passes all tests**.

 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29270/
Test PASSed.

andrewor14 commented:
@CodingCat I never suggested that we grab all cores in spread out mode. The decision of how many cores to give each worker is the same as before. What's different is how we translate those cores into executors. Previously we launch one executor with all the cores given to a worker. Now I am suggesting that we launch multiple executors on the worker, each of which has at most `spark.deploy.maxCoresPerExecutor` cores. Note that if `maxCoresPerExecutor` is not defined, the behavior is the same as the old one, where we just launch one giant executor on the worker with all the cores it has been given.

Does that make sense?

CodingCat commented:
I see.....sorry for the misunderstanding before, I will fix it ASAP

the approach you proposed is quite neat~~

CodingCat committed:
start multiple executors on the worker by rewriting startExeuctor logic

CodingCat committed:
improve the code and docs

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29726/
Test FAILed.

CodingCat committed:
change the description of the parameter in the submit script

CodingCat committed:
recover some unintentional changes

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29728/
Test FAILed.

CodingCat committed:
remove unused var

CodingCat committed:
stylistic fix

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29729/
Test PASSed.

CodingCat commented:
![image](https://cloud.githubusercontent.com/assets/678008/6997678/9d33b010-db93-11e4-88d2-a6e0c225ab88.png)

CodingCat commented:
![image](https://cloud.githubusercontent.com/assets/678008/6997683/a84d083e-db93-11e4-8e02-8363075293f8.png)

CodingCat commented:
@andrewor14  I implement what you proposed, the only difference might be that, in the while condition, we also need to consider the memory constraint on the worker

Thanks for the review

I just uploaded the snapshots


the second picture shows that if we set per executor memory as 6G, the worker with only 10G memory can only afford one executor even it has extra cores

andrewor14 commented on docs/configuration.md:
```diff
@@ -714,6 +714,15 @@ Apart from these, the following properties are also available, and may be useful
+  <td>
+    The maximum number of cores given to the executor. When this parameter is set, Spark will try to 
+    run more than 1 executors on each worker in standalone mode; otherwise, only one executor is 
+    launched on each worker.
```
We should note that this is 1 executor per application. Technically a worker can still run multiple executors if they belong to different applications. I would rephrase this as:

"""
The maximum number of cores given to an executor. When this parameter is set, multiple executors from the same application may run on the same worker, each with cores equal to or fewer than the configured value. Otherwise, at most one executor per application may run on each worker, as that executor will acquire all the worker's cores by default.

This is used in standalone mode only.
"""

andrewor14 replied:
We should note that this is 1 executor per application. Technically a worker can still run multiple executors if they belong to different applications. I would rephrase this as:

"
The maximum number of cores given to an executor. When this parameter is set, multiple executors from the same application may run on the same worker, each with cores equal to or fewer than the configured value. Otherwise, at most one executor per application may run on each worker, as that executor will acquire all the worker's cores by default.

This is used in standalone mode only.
"

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/ApplicationDescription.scala:
```diff
@@ -20,12 +20,13 @@ package org.apache.spark.deploy
     // short name of compression codec used when writing event logs, if any (e.g. lzf)
-    val eventLogCodec: Option[String] = None)
+    val eventLogCodec: Option[String] = None,
+    val maxCorePerExecutor: Option[Int] = None)
```
maxCoresPerExecutor

andrewor14 replied:
as discussed, this would be `coresPerExecutor` instead

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala:
```diff
@@ -480,10 +480,15 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
         | YARN-only:
         |  --driver-cores NUM          Number of cores used by the driver, only in cluster mode
         |                              (Default: 1).
-        |  --executor-cores NUM        Number of cores per executor (Default: 1).
```
please revert these changes

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -374,6 +374,8 @@ object SparkSubmit {
 
       // Other options
+      OptionAssigner(args.executorCores, STANDALONE, ALL_DEPLOY_MODES,
+        sysProp = "spark.deploy.maxCoresPerExecutor"),
```
please revert this change. I don't think we want to confuse `--executor-cores` with the new config

andrewor14 replied:
this will become `spark.executor.cores`

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -524,49 +524,26 @@ private[master] class Master(
-      }
-    }
-
+  private def startExecutorsOnWorkers() {
```
`Unit` return type

andrewor14 replied:
here and everywhere

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -583,31 +560,68 @@ private[master] class Master(
+      app: ApplicationInfo,
+      coresDemand: Int,
+      worker: WorkerInfo): Unit = {
+      if (canUse(app, worker)) {
```
unindent this block by 2 spaces

andrewor14 commented:
@CodingCat Thanks for the latest changes. It is much simpler and I believe it does what we want!

On a separate note, I had an offline discussion with @pwendell about the config semantics. He actually proposes that we configure the number of cores an executor will have exactly, rather than the maximum number cores it could have. Meaning, instead of having `spark.deploy.maxCoresPerExecutor`, we will reuse `spark.executor.cores` as suggested before, but modify the code a little to make sure each executor has exactly N cores instead of at most N cores (where N is the value of `spark.executor.cores`). I will make more suggestions inline to indicate what I mean.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala:
```diff
@@ -480,10 +480,15 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
+        |  --executor-cores NUM        Number of cores to use on each executor. Default:
+        |                              1 in YARN mode; in standalone mode, all cores in a worker
+        |                              allocated to an application will be assigned to a single
+        |                              executor.
```
I would rephrase this as:

Number of cores per executor. (Default: 1 in YARN mode, or all available cores on the worker in standalone mode)

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -583,31 +560,68 @@ private[master] class Master(
+  /**
+   * allocate resources in a certain worker to one or more executors
+   * @param app the info of the application which the executors belong to
+   * @param coresDemand the total number of cores to be allocated to this application
```
More specifically, this should be "cores on this worker to be allocated to this application". I would rename this variable to `coresToAllocate`

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -583,31 +560,68 @@ private[master] class Master(
+      if (canUse(app, worker)) {
+        val memoryPerExecutor = app.desc.memoryPerExecutorMB
+        val maxCoresPerExecutor = app.desc.maxCorePerExecutor.getOrElse(Int.MaxValue)
+        var coresToAssign = coresDemand
```
`var coresLeft = coresToAssign`

andrewor14 commented on docs/configuration.md:
```diff
@@ -714,6 +714,15 @@ Apart from these, the following properties are also available, and may be useful
   </td>
 </tr>
 <tr>
+  <td><code>spark.deploy.maxCoresPerExecutor</code></td>
```
This will be `spark.executor.cores` instead (which is currently not already documented). The description would be something like:
"""
Default: 2 in YARN mode, all the available cores on the worker in standalone mode.

The number of cores to use on each executor. For YARN and standalone mode only.

In standalone mode, setting this parameter allows an application to run multiple executors on the same worker, provided that there are enough cores on that worker. Otherwise, only one executor per application will run on each worker.
"""

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -583,31 +560,68 @@ private[master] class Master(
+      worker: WorkerInfo): Unit = {
+      if (canUse(app, worker)) {
+        val memoryPerExecutor = app.desc.memoryPerExecutorMB
+        val maxCoresPerExecutor = app.desc.maxCorePerExecutor.getOrElse(Int.MaxValue)
```
then this will be
```
val coresPerExecutor = app.desc.coresPerExecutor.getOrElse(coresToAllocate)
```

SparkQA commented:
[Test build #29965 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29965/consoleFull) for   PR 731 at commit [`bc82820`](https://github.com/apache/spark/commit/bc828200402981cd759dd63e04718fe9ea737e06).

CodingCat committed:
change the semantic of coresPerExecutor to exact core number

SparkQA commented:
[Test build #29965 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29965/consoleFull) for   PR 731 at commit [`bc82820`](https://github.com/apache/spark/commit/bc828200402981cd759dd63e04718fe9ea737e06).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29965/
Test FAILed.

CodingCat commented:
@andrewor14 I just updated the patch (the failed test cases are irrelevant to the modification, as nearly all PRs failed on the same reason today)

The modification is a bit more than what you mentioned to ensure that the executor gets exact number of executors, e.g. we need to change this line https://github.com/CodingCat/spark/commit/bc828200402981cd759dd63e04718fe9ea737e06#diff-29dffdccd5a7f4c8b496c293e87c8668L591


the snapshots are following

andrewor14 commented:
retest this please

CodingCat commented:
###Case 1###

this is the command to start the application `bin/spark-shell --executor-cores 2 --total-executor-cores 8 --master spark://127.0.0.1:7077`

we can start multiple executors per worker 

![image](https://cloud.githubusercontent.com/assets/678008/7078714/beda06e8-dee9-11e4-86bc-d9af6319d247.png)



###Case 2###

this is the command to start the application `bin/spark-shell --executor-cores 3 --total-executor-cores 8 --master spark://127.0.0.1:7077`

you can see that, though we have 4 cores on each worker, we can only start 1 executor per worker since we expect the executor expects 3 cores...

![image](https://cloud.githubusercontent.com/assets/678008/7078671/5b85b60a-dee9-11e4-9ab4-c6df350a5e2f.png)

SparkQA commented:
[Test build #29981 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29981/consoleFull) for   PR 731 at commit [`bc82820`](https://github.com/apache/spark/commit/bc828200402981cd759dd63e04718fe9ea737e06).

CodingCat commented:
BTW, I didn't move 'coresPerExecutor' (https://github.com/apache/spark/commit/bc828200402981cd759dd63e04718fe9ea737e06#diff-d3a68f6983a8c3c441b61fe776d77ffdR29) to be close to memoryPerExecutorMB, since it has a default value on that, 

a parameter with the default locating in the middle of the parameter list makes the statement calling that function a bit weird (command does not have a default value)

andrewor14 commented:
Ok, no problem. I will look at this now.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/ApplicationInfo.scala:
```diff
@@ -75,9 +75,9 @@ private[deploy] class ApplicationInfo(
-  ExecutorDesc = {
-    val exec = new ExecutorDesc(newExecutorId(useID), this, worker, cores, desc.memoryPerSlave)
+  private[master] def addExecutor(worker: WorkerInfo, cores: Int, useID: Option[Int] = None):
+      ExecutorDesc = {
```
I appreciate the attempt to fix this, but the style is still wrong. It should be:
```
private[master] def addExecutor(
    worker: WorkerInfo,
    ...): ExecutorDesc = {
  ...
}
```

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -524,49 +524,26 @@ private[master] class Master(
   private def canUse(app: ApplicationInfo, worker: WorkerInfo): Boolean = {
-    worker.memoryFree >= app.desc.memoryPerSlave && !worker.hasExecutor(app)
+    val enoughResources = worker.memoryFree >= app.desc.memoryPerExecutorMB && worker.coresFree > 0
+    val allowToExecute = app.desc.coresPerExecutor.isDefined || !worker.hasExecutor(app)
```
I don't think we need this check. This should already be handled downstream because `worker.coresFree` will be 0 after we launch the first executor. Can you update the javadocs to say:
```
Return whether the worker has enough resources for the app to launch executors on it.
```

andrewor14 replied:
actually, let's just remove this method. We only use it in 1 place and it's small enough that we really don't need it.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -524,49 +524,26 @@ private[master] class Master(
+   * spreadOut mode otherwise packs each app into as few workers as possible until it has assigned
+   * all its cores. User can define spark.deploy.maxCoresPerExecutor per application to
+   * limit the maximum number of cores to allocate to each executor on each worker; if the parameter
+   * is not defined, then only one executor will be launched on a worker.
```
This java doc is verbose and outdated. Can you do the following instead:
```
Schedule executors to be launched on the workers.

There are two modes of launching executors. The first attempts to spread out
an application's executors on as many workers as possible, while the second
does the opposite (i.e. as few workers as possible). The former is usually
better for data locality purposes and is the default.

The number of cores assigned to each executor is configurable. When this
is explicitly set, multiple executors from the same application may be launched
on the worker if the worker has enough cores. Otherwise, each executor grabs
all the cores available on the worker by default, in which case only one executor
per application may be launched on the worker.
```

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -583,31 +560,67 @@ private[master] class Master(
 
-  private def launchExecutor(worker: WorkerInfo, exec: ExecutorDesc) {
+  /**
+   * allocate resources in a certain worker to one or more executors
```
Allocate a worker's resources to one or more executors.

SparkQA commented:
[Test build #29981 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29981/consoleFull) for   PR 731 at commit [`bc82820`](https://github.com/apache/spark/commit/bc828200402981cd759dd63e04718fe9ea737e06).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  case class UnregisterApplication(appId: String)`
  * `class RegexTokenizer extends UnaryTransformer[String, Seq[String], RegexTokenizer] `
  * `    case class Data(weight: Double, mu: Vector, sigma: Matrix)`
  * `    implicit class DslSymbol(sym: Symbol) extends ImplicitAttribute `

 * This patch **removes the following dependencies:**
   * `RoaringBitmap-0.4.5.jar`
   * `activation-1.1.jar`
   * `akka-actor_2.10-2.3.4-spark.jar`
   * `akka-remote_2.10-2.3.4-spark.jar`
   * `akka-slf4j_2.10-2.3.4-spark.jar`
   * `aopalliance-1.0.jar`
   * `arpack_combined_all-0.1.jar`
   * `avro-1.7.6.jar`
   * `breeze-macros_2.10-0.11.1.jar`
   * `breeze_2.10-0.11.1.jar`
   * `chill-java-0.5.0.jar`
   * `chill_2.10-0.5.0.jar`
   * `commons-beanutils-1.7.0.jar`
   * `commons-beanutils-core-1.8.0.jar`
   * `commons-cli-1.2.jar`
   * `commons-codec-1.10.jar`
   * `commons-collections-3.2.1.jar`
   * `commons-compress-1.4.1.jar`
   * `commons-configuration-1.6.jar`
   * `commons-digester-1.8.jar`
   * `commons-httpclient-3.1.jar`
   * `commons-io-2.1.jar`
   * `commons-lang-2.5.jar`
   * `commons-lang3-3.3.2.jar`
   * `commons-math-2.1.jar`
   * `commons-math3-3.1.1.jar`
   * `commons-net-2.2.jar`
   * `compress-lzf-1.0.0.jar`
   * `config-1.2.1.jar`
   * `core-1.1.2.jar`
   * `curator-client-2.4.0.jar`
   * `curator-framework-2.4.0.jar`
   * `curator-recipes-2.4.0.jar`
   * `gmbal-api-only-3.0.0-b023.jar`
   * `grizzly-framework-2.1.2.jar`
   * `grizzly-http-2.1.2.jar`
   * `grizzly-http-server-2.1.2.jar`
   * `grizzly-http-servlet-2.1.2.jar`
   * `grizzly-rcm-2.1.2.jar`
   * `groovy-all-2.3.7.jar`
   * `guava-14.0.1.jar`
   * `guice-3.0.jar`
   * `hadoop-annotations-2.2.0.jar`
   * `hadoop-auth-2.2.0.jar`
   * `hadoop-client-2.2.0.jar`
   * `hadoop-common-2.2.0.jar`
   * `hadoop-hdfs-2.2.0.jar`
   * `hadoop-mapreduce-client-app-2.2.0.jar`
   * `hadoop-mapreduce-client-common-2.2.0.jar`
   * `hadoop-mapreduce-client-core-2.2.0.jar`
   * `hadoop-mapreduce-client-jobclient-2.2.0.jar`
   * `hadoop-mapreduce-client-shuffle-2.2.0.jar`
   * `hadoop-yarn-api-2.2.0.jar`
   * `hadoop-yarn-client-2.2.0.jar`
   * `hadoop-yarn-common-2.2.0.jar`
   * `hadoop-yarn-server-common-2.2.0.jar`
   * `ivy-2.4.0.jar`
   * `jackson-annotations-2.4.0.jar`
   * `jackson-core-2.4.4.jar`
   * `jackson-core-asl-1.8.8.jar`
   * `jackson-databind-2.4.4.jar`
   * `jackson-jaxrs-1.8.8.jar`
   * `jackson-mapper-asl-1.8.8.jar`
   * `jackson-module-scala_2.10-2.4.4.jar`
   * `jackson-xc-1.8.8.jar`
   * `jansi-1.4.jar`
   * `javax.inject-1.jar`
   * `javax.servlet-3.0.0.v201112011016.jar`
   * `javax.servlet-3.1.jar`
   * `javax.servlet-api-3.0.1.jar`
   * `jaxb-api-2.2.2.jar`
   * `jaxb-impl-2.2.3-1.jar`
   * `jcl-over-slf4j-1.7.10.jar`
   * `jersey-client-1.9.jar`
   * `jersey-core-1.9.jar`
   * `jersey-grizzly2-1.9.jar`
   * `jersey-guice-1.9.jar`
   * `jersey-json-1.9.jar`
   * `jersey-server-1.9.jar`
   * `jersey-test-framework-core-1.9.jar`
   * `jersey-test-framework-grizzly2-1.9.jar`
   * `jets3t-0.7.1.jar`
   * `jettison-1.1.jar`
   * `jetty-util-6.1.26.jar`
   * `jline-0.9.94.jar`
   * `jline-2.10.4.jar`
   * `jodd-core-3.6.3.jar`
   * `json4s-ast_2.10-3.2.10.jar`
   * `json4s-core_2.10-3.2.10.jar`
   * `json4s-jackson_2.10-3.2.10.jar`
   * `jsr305-1.3.9.jar`
   * `jtransforms-2.4.0.jar`
   * `jul-to-slf4j-1.7.10.jar`
   * `kryo-2.21.jar`
   * `log4j-1.2.17.jar`
   * `lz4-1.2.0.jar`
   * `management-api-3.0.0-b012.jar`
   * `mesos-0.21.0-shaded-protobuf.jar`
   * `metrics-core-3.1.0.jar`
   * `metrics-graphite-3.1.0.jar`
   * `metrics-json-3.1.0.jar`
   * `metrics-jvm-3.1.0.jar`
   * `minlog-1.2.jar`
   * `netty-3.8.0.Final.jar`
   * `netty-all-4.0.23.Final.jar`
   * `objenesis-1.2.jar`
   * `opencsv-2.3.jar`
   * `oro-2.0.8.jar`
   * `paranamer-2.6.jar`
   * `parquet-column-1.6.0rc3.jar`
   * `parquet-common-1.6.0rc3.jar`
   * `parquet-encoding-1.6.0rc3.jar`
   * `parquet-format-2.2.0-rc1.jar`
   * `parquet-generator-1.6.0rc3.jar`
   * `parquet-hadoop-1.6.0rc3.jar`
   * `parquet-jackson-1.6.0rc3.jar`
   * `protobuf-java-2.4.1.jar`
   * `protobuf-java-2.5.0-spark.jar`
   * `py4j-0.8.2.1.jar`
   * `pyrolite-2.0.1.jar`
   * `quasiquotes_2.10-2.0.1.jar`
   * `reflectasm-1.07-shaded.jar`
   * `scala-compiler-2.10.4.jar`
   * `scala-library-2.10.4.jar`
   * `scala-reflect-2.10.4.jar`
   * `scalap-2.10.4.jar`
   * `scalatest_2.10-2.2.1.jar`
   * `slf4j-api-1.7.10.jar`
   * `slf4j-log4j12-1.7.10.jar`
   * `snappy-java-1.1.1.6.jar`
   * `spark-bagel_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-catalyst_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-core_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-graphx_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-launcher_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-mllib_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-network-common_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-network-shuffle_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-repl_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-sql_2.10-1.4.0-SNAPSHOT.jar`
   * `spark-streaming_2.10-1.4.0-SNAPSHOT.jar`
   * `spire-macros_2.10-0.7.4.jar`
   * `spire_2.10-0.7.4.jar`
   * `stax-api-1.0.1.jar`
   * `stream-2.7.0.jar`
   * `tachyon-0.5.0.jar`
   * `tachyon-client-0.5.0.jar`
   * `uncommons-maths-1.2.2a.jar`
   * `unused-1.0.0.jar`
   * `xmlenc-0.52.jar`
   * `xz-1.0.jar`
   * `zookeeper-3.4.5.jar`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29981/
Test FAILed.

andrewor14 commented:
Weird, retest this please

SparkQA commented:
[Test build #29987 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29987/consoleFull) for   PR 731 at commit [`bc82820`](https://github.com/apache/spark/commit/bc828200402981cd759dd63e04718fe9ea737e06).

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -583,31 +560,67 @@ private[master] class Master(
+   */
+  private def schedule(): Unit = {
+    if (state != RecoveryState.ALIVE) { return }
+    // start in-cluster drivers, they take strict precedence over applications
```
I would just replace this as `// Drivers take strict precedence over executors` and remove the comment in L619, since it doesn't convey any information and doesn't help readability.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -583,31 +560,67 @@ private[master] class Master(
+        }
+      }
+    }
+  }
```
Wait, this method seems to leave out a lot of the existing logic that used to be in `schedule()`. Can you revert this change with the existing comments that you deleted?

andrewor14 commented on docs/configuration.md:
```diff
@@ -714,6 +714,19 @@ Apart from these, the following properties are also available, and may be useful
+  <td><code>spark.executor.cores</code></td>
+  <td>(infinite)</td>
+  <td>
+    Default: 1 in YARN mode, all the available cores on the worker in standalone mode.
```
When I wrote this, I meant to put this phrase into L718, which is really the default column:
```
<td><code>spark.executor.cores</code></td>
<td>1 in YARN mode, all the available cores on the worker in standalone mode.</td>
<td>
  ...
</td>
```

CodingCat replied:
ah, sorry, my fault

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/SparkDeploySchedulerBackend.scala:
```diff
@@ -84,12 +84,11 @@ private[spark] class SparkDeploySchedulerBackend(
-
+    val appDesc = new ApplicationDescription(sc.appName, maxCores, sc.executorMemory,
+      command, appUIAddress, sc.eventLogDir, sc.eventLogCodec,
+      conf.getOption("spark.executor.cores").map(_.toInt))
```
can you put this in a `val coresPerExecutor = conf.getOption(...)` right before this line instead?

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -583,31 +560,67 @@ private[master] class Master(
+    startExecutorsOnWorkers()
+  }
+
+  def launchExecutor(worker: WorkerInfo, exec: ExecutorDesc): Unit =  {
```
1 too many space before {

andrewor14 commented:
@CodingCat This is really close. I think this could be the last real round of reviews before we can merge this. Have you had a chance to test this on a real cluster? It would be good to make sure the scheduling logic is maintained by trying out both with and without the `spreadApp` mode, and launching multiple drivers running at the same time.

CodingCat commented:
sure, I have a 20 cores cluster here, hopefully, I can post the test result here tomorrow

CodingCat committed:
stylistic fixes

SparkQA commented:
[Test build #29992 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29992/consoleFull) for   PR 731 at commit [`45967b4`](https://github.com/apache/spark/commit/45967b4500e0d7332ca82c3511505e1cf586d64c).

CodingCat committed:
remove unused method

CodingCat committed:
revert a change

SparkQA commented:
[Test build #29994 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29994/consoleFull) for   PR 731 at commit [`b8ca561`](https://github.com/apache/spark/commit/b8ca561d48107713f93993729ec4e9aa72375aad).

SparkQA commented:
[Test build #29987 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29987/consoleFull) for   PR 731 at commit [`bc82820`](https://github.com/apache/spark/commit/bc828200402981cd759dd63e04718fe9ea737e06).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29987/
Test PASSed.

SparkQA commented:
[Test build #29992 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29992/consoleFull) for   PR 731 at commit [`45967b4`](https://github.com/apache/spark/commit/45967b4500e0d7332ca82c3511505e1cf586d64c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29992/
Test PASSed.

SparkQA commented:
[Test build #29994 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29994/consoleFull) for   PR 731 at commit [`b8ca561`](https://github.com/apache/spark/commit/b8ca561d48107713f93993729ec4e9aa72375aad).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class FPGrowthModel(JavaModelWrapper):`
  * `class FPGrowth(object):`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29994/
Test PASSed.

CodingCat committed:
avoid unnecessary allocation

SparkQA commented:
[Test build #30016 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30016/consoleFull) for   PR 731 at commit [`940cb42`](https://github.com/apache/spark/commit/940cb4276c7d06f92a7077f32cb2ea748c59a873).

CodingCat commented:
@andrewor14 , I just found that, if we configure exact core number per executor, the current strategy will be incorrect (or say a bit user-unfriendly)

e.g. I have 8 cores, an application would like to use all of them; in spread mode, we will get an array `assigned` as Array(2, 2, 2, 2), if we set --executor-cores as 3, then the application will get 0 core, as we have no allocation no less than 3...

we can say 3 here is not a smart choice, but with this case, the user has to understand how `spread & standalone` works to choose a better number...


shall we go forward with the configuration of the exact number but change the allocation algorithm (bring a much larger patch, like before), or we go back to the max core number configuration?

SparkQA commented:
[Test build #30016 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30016/consoleFull) for   PR 731 at commit [`940cb42`](https://github.com/apache/spark/commit/940cb4276c7d06f92a7077f32cb2ea748c59a873).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30016/
Test PASSed.

CodingCat commented:
after rethinking about this patch, I think it seems to be fine to allocate zero cores in the case I mentioned above, and we just need to filter out those workers if the freeCore in the worker is less than the spark.executor.cores (if it is defined)

SparkQA commented:
[Test build #30097 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30097/consoleFull) for   PR 731 at commit [`da102d4`](https://github.com/apache/spark/commit/da102d42b62c98b50e015bdd705af961361c1063).

CodingCat commented:
Ignore my last comment, I still hold the position that, 

1. if we want to define the exact number of the cores per executor, we need to change the allocation algorithm (bringing a larger patch)

2. otherwise, we go back to define the maxNumberOfCores Per Executor

SparkQA commented:
[Test build #30098 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30098/consoleFull) for   PR 731 at commit [`940cb42`](https://github.com/apache/spark/commit/940cb4276c7d06f92a7077f32cb2ea748c59a873).

SparkQA commented:
[Test build #30098 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30098/consoleFull) for   PR 731 at commit [`940cb42`](https://github.com/apache/spark/commit/940cb4276c7d06f92a7077f32cb2ea748c59a873).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30098/
Test FAILed.

SparkQA commented:
[Test build #30097 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30097/consoleFull) for   PR 731 at commit [`da102d4`](https://github.com/apache/spark/commit/da102d42b62c98b50e015bdd705af961361c1063).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30097/
Test PASSed.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -524,52 +524,25 @@ private[master] class Master(
-   * Can an app use the given worker? True if the worker has enough memory and we haven't already
-   * launched an executor for the app on it (right now the standalone backend doesn't like having
-   * two executors on the same worker).
+   * Schedule executors to be launched on the workers.There are two modes of launching executors.
```
space after `.`

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -524,52 +524,25 @@ private[master] class Master(
+   * Schedule executors to be launched on the workers.There are two modes of launching executors.
+   * The first attempts to spread out an application's executors on as many workers as possible,
+   * while the second does the opposite (i.e. launch them on as few workers as possible). The former
+   * is usually better for data locality purposes and is the default. The number of cores assigned
```
We should split on this sentence to form a new paragraph. Right now it's one huge chunk of text

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -524,52 +524,25 @@ private[master] class Master(
         val usableWorkers = workers.toArray.filter(_.state == WorkerState.ALIVE)
-          .filter(canUse(app, _)).sortBy(_.coresFree).reverse
+          .filter(worker => worker.memoryFree >= app.desc.memoryPerExecutorMB &&
+            worker.coresFree > 0)
```
you technically don't need this check, since we already check in L551
```
if (usableWorkers(pos).coresFree - assigned(pos) > 0)
```

CodingCat replied:
I replaced it with 'worker.coresFree >= app.desc.coresPerExecutor.getOrElse(0)', so that we do not need to run the following allocation algorithm for the case I mentioned above 

> e.g. I have 8 cores, 2 cores per machine, an application would like to use all of them; in spread mode, we will get an array assigned as Array(2, 2, 2, 2), if we set --executor-cores as 3, then the application will get 0 core, as we have no allocation which is no less than 3...

CodingCat replied:
should be `worker.coresFree >= app.desc.coresPerExecutor.getOrElse(1)`

andrewor14 replied:
Again, this predicate is actually not needed because we handle it correctly in the line I pointed out earlier. But not a big deal, we can just leave it.

CodingCat replied:
hmmm....if we remove this, in the case above, the user prefers 3 cores per executor and all workers have at most 2 cores, though we will not allocate anything to the worker, we still generate a `assigned ` array....

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -582,32 +555,63 @@ private[master] class Master(
-            }
-          }
+        for (app <- waitingApps if app.coresLeft > 0 &&
+          worker.memoryFree >= app.desc.memoryPerExecutorMB) {
```
No need to check this again here... we already check this in `allocateWorkerResourceToExecutors` L586

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -582,32 +555,63 @@ private[master] class Master(
         }
       }
     }
+    // start executors
```
remove this comment, as it doesn't convey any information

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/master/Master.scala:
```diff
@@ -582,32 +555,63 @@ private[master] class Master(
+   */
+  private def schedule(): Unit = {
+    if (state != RecoveryState.ALIVE) { return }
+    // start in-cluster drivers, they take strict precedence over applications
```
Can you replace this with
```
// Drivers take strict precedence over executors
```

andrewor14 commented:
@CodingCat We shouldn't have to worry about the case when the user asks for more resources per executors than are available on the cluster. If each machine only has 2 cores the user shouldn't ask for 3 per executor. This holds regardless of whether spread out mode is used, since an executor cannot be "split" across machines. The existing approach is fine.

CodingCat committed:
address the comments

CodingCat committed:
change filter predicate

SparkQA commented:
[Test build #30274 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30274/consoleFull) for   PR 731 at commit [`6dee808`](https://github.com/apache/spark/commit/6dee808d4da7da43805e4ed16fc553f7bc18f494).

andrewor14 commented:
@CodingCat I'm merging this into master. Thanks for keeping this patch open for a long time and patiently reiterating on the reviews. I think the final solution we have here is much simpler than the one we began with.

CodingCat commented:
Hey, @andrewor14 , my pleasure,  many thanks for your patient review

SparkQA commented:
[Test build #30274 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30274/consoleFull) for   PR 731 at commit [`6dee808`](https://github.com/apache/spark/commit/6dee808d4da7da43805e4ed16fc553f7bc18f494).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30274/
Test PASSed.
